### PR TITLE
Optimize index

### PR DESCRIPTION
**Pull request by aawsome**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Optimize creation of new index so that already existing index files are used.
Creation of new index is used at the moment in prune and rebuild-index.

With this pull request the index creation is much faster, especially for remote repositories.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
The change is similar to one part of PR #2340 and the problem is discussed in many issues, e.g. #2162 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by aawsome** [comment]
Seems like the original PR broke some tests with packs in more than one index. This should be fixed now.

**Comment by aawsome** [comment]
While still being confident that this PR can improve the current situation for some users, I now believe that on the long run a refactoring of `rebuild-index` and `prune` (and thus removing `internal/index`) should be favored.

**Comment by irasnyd** [comment]
Hi @aawsome,

I used this PR in conjunction with #2513 during my prune of a large (12M object / ~55TB) AWS S3 backed repository. I don't know how to tell if it was faster or slower than normal, since this was my first prune of this very large repository. As far as I can tell, it worked great. Backups have continued normally without any issue since the repository was pruned (almost 2 weeks ago).

It isn't much feedback, but it's the best I can do. I wanted to give something back, since I had used and tested this PR. Thanks!

**Comment by aawsome** [comment]
@irasnyd 

> I used this PR in conjunction with #2513 during my prune of a large (12M object / ~55TB) AWS S3 backed repository.

Just to clarify: As far as I understood you used `cleanup-index` and `cleanup-packs` from #2513, right?
In this case this PR was not used at all. These two commands do not use `internal/index` as there is already an index data structure in `repository/index` using both data structure multiplies the memory usage and is one of the reasons why `prune` has a memory issue (besides the performance issue due to reading all packs)

**Comment by irasnyd** [comment]
> @irasnyd
> 
> > I used this PR in conjunction with #2513 during my prune of a large (12M object / ~55TB) AWS S3 backed repository.
> 
> Just to clarify: As far as I understood you used `cleanup-index` and `cleanup-packs` from #2513 instead of `prune`, right?

Yes, that is correct. I only used the `cleanup-index` and `cleanup-packs` commands from #2513. I did **not** use `prune`.

> In this case this PR was not used at all. These two commands do not use `internal/index` as there is already an index data structure in `repository/index`. Using both data structure multiplies the memory usage and is one of the reasons why `prune` has a memory issue (besides the performance issue due to reading all packs which can be improved by using this PR)

Apologies for my confusion. I understand better what is going on now. Thanks!

File: internal/index/index.go [hunk 1]
```diff
@@ -76,6 +84,16 @@ func New(ctx context.Context, repo Lister, ignorePacks restic.IDSet, p *restic.P
 				return nil
 			}
 
+			if res, ok := oldIdx.Packs[id]; ok {
```
Comment (MichaelEischer) [inline internal/index/index.go]: This looks like a race condition with the `for res := range outputCh` loop. Two coroutines are working on `idx` and I don't see any synchronization in the AddPack method.

Replacing `repo.ListPack` with a lookup into the old index should avoid that race condition.

File: internal/index/index.go [hunk 2]
```diff
@@ -76,6 +84,16 @@ func New(ctx context.Context, repo Lister, ignorePacks restic.IDSet, p *restic.P
...
+				if err := idx.AddPack(res.ID, size, res.Entries); err != nil {
+					return err
+				}
+				if err := oldIdx.RemovePack(id); err != nil {
```
Comment (MichaelEischer) [inline internal/index/index.go]: Why do you remove the old pack? Is this done to reduce memory usage?

File: internal/index/index.go [hunk 3]
```diff
@@ -255,13 +271,11 @@ func Load(ctx context.Context, repo ListLoader, p *restic.Progress) (*Index, err
 	return index, nil
 }
 
-// AddPack adds a pack to the index. If this pack is already in the index, an
-// error is returned.
+// AddPack adds a pack to the index.
 func (idx *Index) AddPack(id restic.ID, size int64, entries []restic.Blob) error {
 	if _, ok := idx.Packs[id]; ok {
-		return errors.Errorf("pack %v already present in the index", id.Str())
+		debug.Log("pack %v already present in the index", id.Str())
```
Comment (MichaelEischer) [inline internal/index/index.go]: What is the reason for this change? According to my understand of the code it just replaces loading a pack with reusing the old index for that pack. This shouldn't lead to any pack being added twice.

File: internal/index/index.go (hunk 1 above)
Comment (MichaelEischer) [reply internal/index/index.go]: Or better: Just send the pack from the oldIdx over the `outputCh` channel

File: internal/index/index.go (hunk 2 above)
Comment (aawsome) [reply internal/index/index.go]: Yes that was what I was thinking. I actually don't know if the GC is really freeing something in practise, though...

File: internal/index/index.go (hunk 3 above)
Comment (aawsome) [reply internal/index/index.go]: It is not the call in `New` this change is needed for but the call in `Load`. Seems there are some integration tests with index where duplicate packs are present in the index (either within one index file or over different index files)...

**Comment by aawsome** [comment]
I close this PR for two reasons:

- with #2513 there is a functionality to replace `prune` which I assume to be even faster than prune with this functionality. Moreover, this PR is not needed for #2513 
- in #2523 I already proposed a new index implementation which IMO should be enhanced such that it can also be used within #2513 and then 'prune' can be replaced.
